### PR TITLE
Hedera-side fees per user

### LINKED ISSUE #682: Hedera-side fees per user
TLDR: 

- Use ~HFS~/HCS for storing the fee policies for all Hedera native-related operations
    - YAML for fee policies → submit in the topic;
    - validators listen to the topic + chunk logic
    - validators load the YAML in memory
    - similar to bridge config
    - HBAR → EVM; HTS → EVM
    - sender account is used for identification of the fee policy

HFS ruled out for storing fee policies as there is no way to get file contents from the mirror-node, which means it would cost network fees to read files. 

<hr>

1. Proposed fee policy model
```yaml
type: "flat | percentage | tiered_percentage" 
amount: 10000 # in case of flat fee
percentage: 500 # in case of percentage fee
tiers: # in case of tiered fee
   - lte: 1000 # less-than-or-equal range of amount of token where the tier is applied
     percentage: 20000 # 20.00%
   - lte: 10000 
     percentage: 10000 # 10.00%
```

2. Create a topic that will contain the fee policies for all clients requiring them.
Proposed YAML structure: 
```yaml
policies:
  <chain ID>: 
    <token ID | address>:
       <sender acc ID | address>: 
          type: flat
          amount: 500  
```
 The policies should be sent to the topic in chunks utilizing [ConsensusMessageChunkInfo](https://docs.hedera.com/guides/docs/hedera-api/consensus-service/consensussubmitmessage#consensusmessagechunkinfo), assembled on start of the validator and loaded in memory. 

⚠️ It is important to validate the policies, e.g. 
 - NFT assets cannot have a percentage or tiered base fee policy. 
 - There should be only 1 policy per asset per account. 

3. Whenever a transfer is initiated, look up this fee policy configuration for an alternate fee policy for the originator account ID of the transaction. If such exists, use it instead.

[User Stories](https://www.notion.so/limechain/Fee-Per-User-User-Stories-f7fc1ebd1cdc45a98c8f50b74ecc968f)

### PR DESCRIPTION
**Pull request by trayanm**
Add fee policy handler to handle topic and policies per legal entity
Add fee policy implementations for flat fee, percentage fee, fee per token 
Change fee calculation to use fee policies
Change proto to hold fee
Change message processing to handle message with fee 
Change router contract proxy
Change e2e tests

**Detailed description**:

**Which issue(s) this PR fixes**:
Fixes #682

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

File: app/domain/service/contracts.go [hunk 1]
```diff
@@ -54,4 +54,13 @@ type Contracts interface {
...
 	WatchLockEventLogs(opts *bind.WatchOpts, sink chan<- *abi.RouterLock) (event.Subscription, error)
+	// FeeAmountFor returns fee data for specific bridge operation
+	FeeAmountFor(_targetChain *big.Int, _userAddress common.Address, _tokenAddress common.Address, _amount *big.Int) (*big.Int, error)
+
+	TokenFeeData(_token common.Address) (struct {
+		ServiceFeePercentage *big.Int
+		FeesAccrued          *big.Int
+		PreviousAccrued      *big.Int
+		Accumulator          *big.Int
+	}, error)
```
Comment (rokn) [inline app/domain/service/contracts.go:65]: Do we need these params to be underscored for some reason or is coming from somewhere else?

File: app/domain/service/fee.go [hunk 2]
```diff
@@ -20,4 +20,6 @@ package service
 type Fee interface {
 	// CalculateFee calculates the fee and remainder of a given amount, based on a specified token fee percentage
 	CalculateFee(token string, amount int64) (fee, remainder int64)
+
+	CalculatePercentageFee(amount int64, feePercentage int64) (fee, remainder int64)
```
Comment (rokn) [inline app/domain/service/fee.go:24]: nit: docs

File: app/domain/service/messages.go [hunk 3]
```diff
@@ -32,6 +32,8 @@ type Messages interface {
 	ProcessSignature(transferID, signature string, targetChainId uint64, timestamp int64, authMsg []byte) error
 	// SignFungibleMessage signs a Fungible message based on Transfer
 	SignFungibleMessage(transfer payload.Transfer) ([]byte, error)
+	// SignFungibleMessageWithFee signs a Fungible message based on Transfer
```
Comment (rokn) [inline app/domain/service/messages.go:35]: Add info that this includes the fee from the message

File: app/model/auth-message/auth-message.go [hunk 4]
```diff
@@ -43,9 +44,46 @@ func EncodeFungibleBytesFrom(sourceChainId, targetChainId uint64, txId, asset, r
...
+	if err != nil {
+		return nil, err
+	}
+
+	return keccak(bytesToHash), nil
+}
+
+// EncodeFungibleBytesFromWithFee returns the array of bytes representing an
+// authorisation ERC-20 Mint signature ready to be signed by EVM Private Key
+func EncodeFungibleBytesFromWithFee(sourceChainId, targetChainId uint64, txId, asset, receiverEthAddress, amount string, fee string) ([]byte, error) {
```
Comment (rokn) [inline app/model/auth-message/auth-message.go:57]: Can we add tests here

File: app/model/auth-message/auth-message.go [hunk 5]
```diff
@@ -154,6 +192,46 @@ func generateFungibleArguments() (abi.Arguments, error) {
 		}}, nil
 }
 
+func generateFungibleArgumentsWithFee() (abi.Arguments, error) {
```
Comment (rokn) [inline app/model/auth-message/auth-message.go:195]: This can reuse the `func generateFungibleArguments()` by appending the last argument for the fee to the result.

File: app/model/auth-message/auth-message_test.go [hunk 6]
```diff
@@ -1,8 +1,9 @@
 package auth_message
 
 import (
-	"github.com/stretchr/testify/assert"
 	"testing"
+
+	"github.com/stretchr/testify/assert"
```
Comment (rokn) [inline app/model/auth-message/auth-message_test.go:6]: nit: we should skip committing formatting changes IMO

File: app/model/fee-policy/fee_policy_interface.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fee_policy
+
+type FeePolicyInterface interface {
```
Comment (rokn) [inline app/model/fee-policy/fee_policy_interface.go:19]:
```suggestion
type FeePolicy interface {
```
We should try to keep to the code style established in the codebase

File: app/model/fee-policy/flat_fee_per_token_policy.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fee_policy
+
+type FlatFeePerTokenPolicy struct {
+	Networks  []uint64
+	TokenFees map[string]int64 // { token: "0.0.4564", value: 7 }
```
Comment (rokn) [inline app/model/fee-policy/flat_fee_per_token_policy.go:21]: nit: I think this comment will be better by showing the exact mapping:
`{ "0.0.4564":  7 }` otherwise it looks a bit like it's some kind of strange map which has a field token

File: app/model/fee-policy/flat_fee_per_token_policy.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	pairs := parsingValue.([]interface{})
+
+	for _, item := range pairs {
+		token, foundToken := getInterfaceValue(item, "token")
+		value, foundValue := getInterfaceValue(item, "value")
+
+		if foundToken && foundValue {
+			result.TokenFees[token.(string)] = (int64(value.(int)))
+		}
+	}
```
Comment (rokn) [inline app/model/fee-policy/flat_fee_per_token_policy.go:39]: Do we need to error out here if the pairs don't contain token or value?

File: app/model/fee-policy/flat_fee_per_token_policy.go [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+	return result
+}
+
+func (policy *FlatFeePerTokenPolicy) FeeAmountFor(networkId uint64, token string, amount int64) (feeAmount int64, exist bool) {
```
Comment (rokn) [inline app/model/fee-policy/flat_fee_per_token_policy.go:44]: nit: I think we can change the amount to _ since we are not using it.

File: app/model/fee-policy/flat_fee_per_token_policy.go [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+	var found bool = networkFound(policy.Networks, networkId)
```
Comment (rokn) [inline app/model/fee-policy/flat_fee_per_token_policy.go:45]: nit: type is not needed here

File: app/model/fee-policy/helpers.go [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+		if eleKey.(string) == key {
+			return eleValue, true
+		}
+	}
+
+	return nil, false
+}
+
+func networkFound(networks []uint64, networkId uint64) bool {
+	if networks != nil && networkId > 0 {
```
Comment (rokn) [inline app/model/fee-policy/helpers.go:35]: These checks aren't necessary. If `networks` is nil the for cycle won't start. If the `networkId == 0` then it just won't find it. If we want to leave the `networkId` check we should convert it to an early exit:
`if networkId == 0 { return false }`

File: app/model/fee-policy/helpers.go [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+		for _, ele := range networks {
+			if ele == networkId {
+				return true
+			}
+		}
+	}
+
+	return true
```
Comment (rokn) [inline app/model/fee-policy/helpers.go:43]:
```suggestion
	return false
```

File: app/model/fee-policy/percentage_fee_policy.go [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+type PercentageFeePolicy struct {
+	Networks []uint64
+	Value    int64
+}
+
+func ParseNewPercentageFeePolicy(networks []uint64, parsingValue interface{}) *PercentageFeePolicy {
+	return &PercentageFeePolicy{
+		Networks: networks,
+		Value:    (int64(parsingValue.(int))),
```
Comment (rokn) [inline app/model/fee-policy/percentage_fee_policy.go:29]: nit: can we directly parse to int64?

File: app/model/fee-policy/percentage_fee_policy.go [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+	}
+}
+
+func (policy *PercentageFeePolicy) FeeAmountFor(networkId uint64, token string, amount int64) (feeAmount int64, exist bool) {
+	var found bool = networkFound(policy.Networks, networkId)
```
Comment (rokn) [inline app/model/fee-policy/percentage_fee_policy.go:34]: nit: same here for the bool type

File: app/process/handler/message-submission/handler_test.go [hunk 16]
```diff
@@ -79,9 +79,16 @@ var (
 		Amount:        transferRecord.Amount,
 		Signature:     "signature",
 	}
-	authMsgBytes, _ = auth_message.EncodeFungibleBytesFrom(transferRecord.SourceChainID, transferRecord.TargetChainID, transferRecord.TransactionID, transferRecord.TargetAsset, transferRecord.Receiver, transferRecord.Amount)
-	date            = time.Date(2001, time.June, 1, 1, 1, 1, 1, time.UTC)
-	txId            = &hedera.TransactionID{
+	authMsgBytes, _ = auth_message.EncodeFungibleBytesFrom(
```
Comment (rokn) [inline app/process/handler/message-submission/handler_test.go:82]: nit: formatting change

File: app/process/handler/message/handler.go [hunk 17]
```diff
@@ -118,11 +119,21 @@ func (cmh Handler) handleFungibleSignatureMessage(tsm *proto.TopicEthSignatureMe
...
 	}
 
-	// Parse incoming message
-	authMsgBytes, err := auth_message.EncodeFungibleBytesFrom(tsm.SourceChainId, tsm.TargetChainId, tsm.TransferID, tsm.Asset, tsm.Recipient, tsm.Amount)
-	if err != nil {
-		cmh.logger.Errorf("[%s] - Failed to encode the authorisation signature. Error: [%s]", tsm.TransferID, err)
-		return
+	var authMsgBytes []byte
+
+	// if Topic Signature Message contains Fee - try to parse it first
```
Comment (rokn) [inline app/process/handler/message/handler.go:124]: nit: 
```suggestion
	// if Topic Signature Message contains Fee - generate one which contains fee as well
```

File: app/process/watcher/fee-policy-config/watcher.go [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+func (watcher *Watcher) watchIteration() {
+	watcher.logger.Infof("Checking for new Fee Policy Config ...")
+	_, err := watcher.feePolicyHandler.ProcessLatestFeePolicyConfig(watcher.feePolicyTopicID)
+
+	if err != nil {
+		watcher.logger.Errorf(err.Error())
+	} else {
+		watcher.logger.Infof("Fee Policy Config finished successfully!")
```
Comment (rokn) [inline app/process/watcher/fee-policy-config/watcher.go:64]: nit: can we move this log to the service to keep it the same as the bridge config service

File: app/router/fees/fees.go [hunk 19]
```diff
@@ -29,9 +31,15 @@ import (
 
 const Route = "/fees"
 
-func NewRouter(pricingService service.Pricing) http.Handler {
+type feePercentageInfo struct {
+	Amount        int64 `json:"amount"`
+	MaxPercentage int64 `json:"maxPercentage"`
+}
```
Comment (rokn) [inline app/router/fees/fees.go:37]: This doesn't seem to be used anywhere

File: app/domain/service/contracts.go (hunk 1 above)
Comment (trayanm) [reply app/domain/service/contracts.go:65]: It is copy/paste from router contract proxy in `app/clients/evm/contracts/router/diamond-router.go`. It is easier to change the definition here if the proxy is changed.
I will add the comment above `TokenFeeData `

File: app/services/fee/calculator/calculator.go [hunk 20]
```diff
@@ -50,7 +50,13 @@ func New(feePercentages map[string]int64) *Service {
...
 // CalculateFee calculates the fee and remainder of a given token and amount
 func (s Service) CalculateFee(token string, amount int64) (fee, remainder int64) {
-	fee = amount * s.feePercentages[token] / constants.FeeMaxPercentage
+	feePercentage := s.feePercentages[token]
+
+	return s.CalculatePercentageFee(amount, feePercentage)
+}
+
+func (s Service) CalculatePercentageFee(amount int64, feePercentage int64) (fee, remainder int64) {
+	fee = amount * feePercentage / constants.FeeMaxPercentage
```
Comment (rokn) [inline app/services/fee/calculator/calculator.go:59]: Can we add tests here for the new function

File: e2e/e2e_test.go [hunk 21]
```diff
@@ -532,7 +540,7 @@ func Test_E2E_Hedera_EVM_Native_Token(t *testing.T) {
 	chainId := setupEnv.Scenario.FirstEvmChainId
 	evm := setupEnv.Clients.EVM[chainId]
 	memo := fmt.Sprintf("%d-%s", chainId, evm.Receiver.String())
-	unlockAmount := int64(10) // Amount, which converted to 18 decimals is 100000000000 (100 gwei)
+	unlockAmount := int64(100) // Amount, which converted to 18 decimals is 100000000000 (100 gwei)
```
Comment (rokn) [inline e2e/e2e_test.go:543]: Is this change needed? If so, update the comment please

File: e2e/e2e_test.go [hunk 22]
```diff
@@ -637,7 +658,7 @@ func Test_E2E_Hedera_EVM_Native_Token(t *testing.T) {
 
 // Test_EVM_Native_to_EVM_Token recreates a real life situation of a user who wants to bridge an EVM native token to another EVM chain.
 func Test_EVM_Native_to_EVM_Token(t *testing.T) {
-	amount := int64(1000000000000) // 1000 gwei
+	amount := int64(1000000000000000) // 1000 gwei
```
Comment (rokn) [inline e2e/e2e_test.go:661]: nit: here as well if needed, update the comment

File: scripts/bridge/setup/setup.go [hunk 23]
```diff
@@ -158,7 +159,7 @@ func ValidateArguments(privateKey *string, accountID *string, adminKey *string,
 func cryptoCreate(client *hedera.Client, result *DeployResult) (hedera.PrivateKey, error) {
 	privateKey, _ := hedera.PrivateKeyGenerateEd25519()
 	fmt.Printf("Hedera Private Key: %v\n", privateKey.String())
-	fmt.Printf("Hederea Public Key: %v\n", privateKey.PublicKey().String())
+	fmt.Printf("Hedera Public Key: %v\n", privateKey.PublicKey().String())
```
Comment (rokn) [inline scripts/bridge/setup/setup.go:162]: Thank you! This was driving me nuts

File: test/mocks/client/diamond_router_client_mock.go [hunk 24]
```diff
@@ -106,3 +106,13 @@ func (m *MockDiamondRouter) Erc721Payment(opts *bind.CallOpts, _erc721 common.Ad
...
+
+func (m *MockDiamondRouter) FeeAmountFor(opts *bind.CallOpts, _targetChain *big.Int, _userAddress common.Address, _tokenAddress common.Address, _amount *big.Int) (*big.Int, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockDiamondRouter) ServiceFeePrecision(opts *bind.CallOpts) (*big.Int, error) {
+	// TODO implement me
+	panic("implement me")
+}
```
Comment (rokn) [inline test/mocks/client/diamond_router_client_mock.go:118]: Do we need implementations here?

File: app/domain/service/contracts.go (hunk 1 above)
Comment (rokn) [reply app/domain/service/contracts.go:65]: Yeah I see that there they are underscored but here all of the other functions are without underscores so I think it will be best to keep them without them here.

File: app/model/auth-message/auth-message_test.go (hunk 6 above)
Comment (trayanm) [reply app/model/auth-message/auth-message_test.go:6]: imports are rearranged by VS  Code (with default settings) on save

Comment (rokn) [reply app/model/auth-message/auth-message_test.go:6]: Yeah I know that's the cause it's just adds a bit of overhead to the PR. I'm okay with merging them it's just something of note for future PR's if possible

File: test/mocks/client/diamond_router_client_mock.go (hunk 24 above)
Comment (trayanm) [reply test/mocks/client/diamond_router_client_mock.go:118]: the methods are required because of `DiamondRouter interface`.
I will mock the call

File: app/model/fee-policy/helpers.go (hunk 12 above)
Comment (trayanm) [reply app/model/fee-policy/helpers.go:35]: The method name is changed to `networkAllowed` be more clear. Actually the function is checking for restriction (true by default).

File: app/services/contracts/service.go [hunk 25]
```diff
@@ -136,6 +136,19 @@ func (bsc *Service) getMembers() ([]string, error) {
...
+func (bsc *Service) FeeAmountFor(_targetChain *big.Int, _userAddress common.Address, _tokenAddress common.Address, _amount *big.Int) (*big.Int, error) {
+	return bsc.contract.FeeAmountFor(nil, _targetChain, _userAddress, _tokenAddress, _amount)
+}
+
+func (bsc *Service) TokenFeeData(_token common.Address) (struct {
+	ServiceFeePercentage *big.Int
+	FeesAccrued          *big.Int
+	PreviousAccrued      *big.Int
+	Accumulator          *big.Int
+}, error) {
```
[the remark this answers is not in the document]
Comment (trayanm) [reply app/services/contracts/service.go:148]: The method is implementation of `DiamondRouter interface`, where it is defined with the same structure.

File: app/model/auth-message/auth-message.go [hunk 26]
```diff
@@ -17,20 +17,22 @@
...
 
 import (
+	"math/big"
+
 	"github.com/ethereum/go-ethereum/accounts/abi"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/crypto"
-	"github.com/limechain/hedera-eth-bridge-validator/app/helper/big-numbers"
-	"math/big"
+	big_numbers "github.com/limechain/hedera-eth-bridge-validator/app/helper/big-numbers"
```
Comment (rokn) [inline app/model/auth-message/auth-message.go:25]: Why is this rename needed?

File: app/router/fees/fees.go [hunk 27]
```diff
@@ -48,3 +52,40 @@ func feesNftResponse(pricingService service.Pricing) func(w http.ResponseWriter,
 		render.JSON(w, r, res)
 	}
 }
+
+func calculateForResponse(feeService service.Fee, feePolicyHandler service.FeePolicyHandler) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		targetChainIdStr := r.URL.Query().Get("chain")
```
Comment (rokn) [inline app/router/fees/fees.go:58]: I think targetChain would be a more descriptive parameter name here. We are setting the variable to this why not the parameter as well

File: docs/fee-policies.md [hunk 28]
```diff
@@ -0,0 +1,96 @@
+# Fee policies
+
+Fee policies provides special fees to be applied for users from specific `Legal Entity`. Usually the fee policy will apply lesser fee amount over the bridge transaction.
```
Comment (rokn) [inline docs/fee-policies.md:3]:
```suggestion
Fee policies provide special fees to be applied for users from specific `Legal Entity`. Usually the fee policy will apply a lesser fee for bridge transactions.
```

File: docs/fee-policies.md [hunk 29]
```diff
@@ -0,0 +1,96 @@
...
+In order to enable fee policies to the bridge - submit _Topic_ and with this _Topic Id_ in `config/bridge.yml` (See [Configuration](configuration.md), field `bridge.config_topic_id`).
```
Comment (rokn) [inline docs/fee-policies.md:4]: I'm not sure what is meant by this. Probably something along the lines of:
`submit a topic message with the config to the topic id from the config.`

File: docs/fee-policies.md [hunk 30]
```diff
@@ -0,0 +1,96 @@
...
+In order to enable fee policies to the bridge - submit _Topic_ and with this _Topic Id_ in `config/bridge.yml` (See [Configuration](configuration.md), field `bridge.config_topic_id`). 
+
+## Terminology
+During this document and in the code - the following terminology and meaning are used:
+* `Legal Entity` - Legal entity for which the `Fee Policy` is applied. The actual legal entity is used only for logically separation. Subject to the `Fee Policy` are the defined list of `User Address` items.
+* `User Address` - _EVM Wallet Address_ or _Hedera Account Id_ to which the `Fee Policy` will be applied.
+* `Token Address` - _EVM Token Address_ or _Hedera Token Id_ used in bridge transaction for which the `Fee Policy` will be applied.
+* `Target Network` - Network Id target of the bridge transaction, for which the `Fee Policy` will be applied.
+* `Fee Policy` - Fee policy definition, describing the actual fee calculation. The fee policy is an implementation of interface `FeePolicyInterface` in `app/model/fee-policy/fee_policy_interface.go`.
```
Comment (rokn) [inline docs/fee-policies.md:12]:
```suggestion
* `Fee Policy` - Fee policy definition, describing the actual fee calculation. The fee policy is an implementation of interface `FeePolicy` in `app/model/fee-policy/fee_policy.go`.
```

File: docs/fee-policies.md [hunk 31]
```diff
@@ -0,0 +1,96 @@
...
+* `Fee Policy Type` - The type describes the calculation logic of the fee. Possible values are `flat`, `percentage`, `flat_per_token`.
+
+## Fee Policy Config structure
+Available `Fee Policy Type` items are:
+* Flat - code usage `constants.FeePolicyTypeFlat` - `Fee Policy` with specific flat fee without other restrictions.
+* Percentage - code usage `constants.FeePolicyTypePercentage` - `Fee Policy` with specific percentage fee without other restrictions
+* Flat Per Token - code usage `constants.FeePolicyTypeFlatPerToken` - `Fee Policy` with specific flat fee per tokens
```
Comment (rokn) [inline docs/fee-policies.md:19]:
```suggestion
* Flat Per Token - code usage `constants.FeePolicyTypeFlatPerToken` - `Fee Policy` with specific flat fee per token
```

File: app/model/auth-message/auth-message.go (hunk 26 above)
Comment (trayanm) [reply app/model/auth-message/auth-message.go:25]: The rename is done by VS Code on safe.

File: app/model/auth-message/auth-message.go [hunk 32]
```diff
@@ -43,9 +45,46 @@ func EncodeFungibleBytesFrom(sourceChainId, targetChainId uint64, txId, asset, r
...
+	return keccak(bytesToHash), nil
+}
+
+// EncodeFungibleBytesFromWithFee returns the array of bytes representing an
+// authorisation ERC-20 Mint signature ready to be signed by EVM Private Key
+func EncodeFungibleBytesFromWithFee(sourceChainId, targetChainId uint64, txId, asset, receiverEthAddress, amount string, fee string) ([]byte, error) {
+
+	args, err := generateFungibleArguments(true)
+	if err != nil {
+		return nil, err
```
Comment (svetlio8) [inline app/model/auth-message/auth-message.go:62]: unit tests?

File: app/model/fee-policy/flat_fee_per_token_policy.go [hunk 33]
```diff
@@ -0,0 +1,74 @@
+/*
```
Comment (svetlio8) [inline app/model/fee-policy/flat_fee_per_token_policy.go:1]: Could you add unit tests for the new fee-policy folder?

File: app/process/watcher/fee-policy-config/watcher.go [hunk 34]
```diff
@@ -0,0 +1,66 @@
...
+
+	"github.com/hashgraph/hedera-sdk-go/v2"
+	qi "github.com/limechain/hedera-eth-bridge-validator/app/domain/queue"
+	"github.com/limechain/hedera-eth-bridge-validator/app/domain/service"
+	"github.com/limechain/hedera-eth-bridge-validator/config"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	sleepTime = 10 * time.Minute
```
Comment (svetlio8) [inline app/process/watcher/fee-policy-config/watcher.go:30]: Please add it to docs or move it to the config

File: app/services/fee-policy/service.go [hunk 35]
```diff
@@ -0,0 +1,327 @@
...
+	return &Service{
+		mirrorNode:        mirrorNode,
+		queryMaxLimit:     mirrorNode.QueryMaxLimit(),
+		queryDefaultLimit: mirrorNode.QueryDefaultLimit(),
+		config:            cfg,
+		logger:            config.GetLoggerFor("Fee Policy Config Service"),
+	}
+}
+
+func (s *Service) FeeAmountFor(networkId uint64, account string, token string, amount int64) (feeAmount int64, exist bool) {
```
Comment (svetlio8) [inline app/services/fee-policy/service.go:63]: unit tests?

File: test/mocks/service/fee_policy_handler_mock.go [hunk 36]
```diff
@@ -0,0 +1,37 @@
...
+	mock.Mock
+}
+
+func (m *MockFeePolicyHandler) ProcessLatestConfig(topicID hedera.TopicID) (*parser.FeePolicy, error) {
+	args := m.Called(topicID)
+	return args[0].(*parser.FeePolicy), args.Error(1)
+}
+
+func (m *MockFeePolicyHandler) FeeAmountFor(networkId uint64, account string, token string, amount int64) (feeAmount int64, exist bool) {
+	// TODO implement me
```
Comment (svetlio8) [inline test/mocks/service/fee_policy_handler_mock.go:35]: why in todo?

File: test/mocks/service/fee_service_mock.go [hunk 37]
```diff
@@ -26,3 +26,8 @@ func (mfs *MockFeeService) CalculateFee(token string, amount int64) (fee, remain
 	args := mfs.Called(token, amount)
 	return args.Get(0).(int64), args.Get(1).(int64)
 }
+
+func (mfs *MockFeeService) CalculatePercentageFee(amount int64, feePercentage int64) (fee, remainder int64) {
+	// TODO implement me
+	panic("implement me")
```
Comment (svetlio8) [inline test/mocks/service/fee_service_mock.go:32]: why in todo?

File: test/mocks/bridge-provider.go [hunk 38]
```diff
@@ -17,20 +17,31 @@
...
 	mock.Mock
 }
 
+func (m *MockBridgeContract) TokenFeeData(token common.Address) (struct {
+	ServiceFeePercentage *big.Int
+	FeesAccrued          *big.Int
+	PreviousAccrued      *big.Int
+	Accumulator          *big.Int
+}, error) {
+	// TODO implement me
```
[the remark this answers is not in the document]
Comment (trayanm) [reply test/mocks/bridge-provider.go:41]: The method is generated by GoLand IDE to respect Contracts interface. Currently body is not necessary. The TODO comment is already removed

File: test/mocks/service/fee_policy_handler_mock.go (hunk 36 above)
Comment (trayanm) [reply test/mocks/service/fee_policy_handler_mock.go:35]: The method is generated by GoLand IDE to respect FeePolicyHandler interface. Currently body is not necessary. The TODO comment is already removed

File: test/mocks/service/fee_service_mock.go (hunk 37 above)
Comment (trayanm) [reply test/mocks/service/fee_service_mock.go:32]: The method is generated by GoLand IDE to respect Fee interface. Currently body is not necessary. The TODO comment is already removed

**Review by svetlio8 — DISMISSED** [review]

**Review by rokn — DISMISSED** [review]

**Review by rokn — DISMISSED** [review]

**Review by Coiling-Dragon — APPROVED** [review]

**Review by rokn — APPROVED** [review]
